capi/geos_c.h.in

```diff
@@ -306,19 +306,56 @@ typedef int (*GEOSTransformXYCallback)(
 typedef void (GEOSInterruptCallback)(void);
 
 /**
-* Register a function to be called when processing is interrupted.
+* Callback function for use in interruption. The callback will be invoked at each
+* possible interruption point and can be used to request interruption.
+*
+* \see GEOS_interruptRegisterThreadCallback
+* \see GEOS_interruptThread
+*/
+typedef void (GEOSInterruptThreadCallback)(void*);
+
+/**
+* Register a function to be called when a possible interruption point is reached
+* on any thread. The function may be used to request interruption.
+*
 * \param cb Callback function to invoke
-* \return the previously configured callback
+* \return the previously registered callback, or NULL
 * \see GEOSInterruptCallback
+* \see GEOSContext_setInterruptCallback_r
 */
 extern GEOSInterruptCallback GEOS_DLL *GEOS_interruptRegisterCallback(
     GEOSInterruptCallback* cb);
 
 /**
-* Request safe interruption of operations
+* Register a function to be called when a possible interruption point is reached
+* in code executed in the specified context. The function can interrupt the
+* thread if desired by calling GEOS_interruptThread.
+*
+* \param extHandle the context returned by \ref GEOS_init_r.
+* \param cb Callback function to invoke
+* \param userData optional data to be pe provided as argument to callback
+* \return the previously registered callback, or NULL
+* \see GEOSInterruptThreadCallback
+*/
+extern GEOSInterruptThreadCallback GEOS_DLL *GEOSContext_setInterruptCallback_r(
+        GEOSContextHandle_t extHandle,
+        GEOSInterruptThreadCallback* cb,
+        void* userData);
```

Comment:
@dbaston what do you think about making this callback return a value that indicates whether an interrupt is desired and eliminating `GEOS_interruptThread()`?